### DISCUSSION TITLE
The future of The Composable Environment

### DISCUSSION THREAD

**Discussion by maximkrouk** [tgrapperon/swift-composable-environment#13] [General]
A place to discuss the future of the library, API improvements etc.

**Comment by maximkrouk** [comment]
I guess the general approach is very convenient and it's even more then TCA extension lib, probably even other projects (not using TCA) can benefit from TCE, so maybe it worth removing TCA dependency for such projects.

I see 2 ways, the first one is to simply create a separate branch, the second is to remove it completely in `0.6.0`, because using `@DerivedEnvironment` property wrappers (which are lazy now) is probably more efficient than custom pullbacks and also currently it's the only way to use the environment with recursive reducers (maybe it will work, because my implementation of recursive reducers is based on lazy(fying? 😅) the recursion, but I didn't tried it, so I'm not sure. so I use the environment argument in pullbacks anyway).

There is also some wired type inference bug in the pullback's environment property, so that derived environment keyPath cannot be treated as function so I use `get` function from the prelude to transform the keyPath like

```swift
let reducer = Reducer.combine(
  derivedReducer.pullback(
    state: \State.derived,
    action: /Action.derived,
    environment: get(\Environment.derived)
  ),
  ...
)
```

**Reply by tgrapperon** [reply]
The `KeyPath` issue looks like a regression. I'll see what I can do.

**Comment by maximkrouk** [comment]
@tgrapperon

**Reply by tgrapperon** [reply]
Hi, @maximkrouk, sorry for not getting back to you earlier.

I'm not super interested in maintaining a general-purpose dependency management library, so I think I'll keep TCA as a requirement for now. I guess there are already dependency management libraries using some similar SwiftUI-like API's, and I'm not knowledgeable enough with the general picture/architectures to handle such a task.
By keeping the project tied to TCA, we can adapt if the architecture changes its implementation. For example, the way dependencies can be overridden at each node is not super streamlined, with classes and invalidation, but this is what's ultimately working with TCA in the way it's currently implemented. This may change if TCA proto branch lands at some point (superseding this library I guess, but with a similar API), and we may adapt accordingly along the way. 

But feel free to fork the library and experiment with some "Core" component, and a TCA specialized `ComposableEnvironment` that will depend on it.

**Reply by maximkrouk** [reply]
I'm currently working on a project that unfortunately doesn't use TCA 😢, so I created a fork without TCA dependency
https://github.com/capturecontext/swift-composable-environment
